### ISSUE TITLE
Pluggable Congestion Control

### ISSUE THREAD

**Issue by aboba** [w3c/webtransport#365]
QUIC congestion control algorithms such as BBRv1 have issues when WebTransport is used for realtime communications.  For example, there is no equivalent to the probing supported in WebRTC implementations, which allows for rapid rampup in encoder rate when congestion is resolved. 

The problem could be solved by allowing alternate congestion control mechanisms to be specified when the WebTransport is constructed.

**Comment by LPardue** [comment]
> default congestion control for QUIC

Can you clarify this a bit? While some types of QUIC frames are subject to congestion control, there is no default algorithm in QUIC (a reno-inspired congestion control algorithm is provided as an example, it is not required to implement that algorithm and many choose something like cubic or BBR).

Is the request to exempt some type of frames (e.g. STREAM or DATAGRAM) from the transport-layer congestion controller?

**Comment by aboba** [comment]
The [IETF RMCAT WG](https://datatracker.ietf.org/wg/rmcat/about/) developed congestion control algorithms optimized for use with realtime communications.  The algorithms developed were published as [experimental RFCs](https://datatracker.ietf.org/wg/rmcat/documents/). 

The proposal would be to allow selection of one of these algorithms (or an alternative, developed with a similar goal). 

In particular, the algorithms developed in RMCAT (or alternatives implemented in WebRTC.org source) seek to avoid building queues in bottleneck devices by incorporating delay metrics.  They also are capable of probing the network to determine when simulcast or SVC layers can be added back after congestive losses cause the layers to be dropped.

**Comment by LPardue** [comment]
Thanks @aboba, that's helpful. Would this proposal require the browser to implement those CCs on behalf of the API user? If so, are there cases where applications above the API want to run their own CC  (which sounds flexible but also like a footgun).

**Comment by aboba** [comment]
It is currently not possible for realtime communications applications to implement the kind of CC algorithms specified in RMCAT on their own. Firstly, the information necessary to do this is not available (WebTransport does not expose statistics, such as information available to the QUIC implementation.  Also, an application cannot implement multiplicative increase on top of a QUIC congestion control algorithm designed to implement Multiplicate Decrease/Additive Increase. 

Note that this problem has been understood from the start (e.g. it was raised in the original design of RTCQuicTransport), but has fallen through the cracks.

**Comment by LPardue** [comment]
Makes sense! Would this work for WebTransport over HTTP/2?

**Comment by aboba** [comment]
Getting statistics (such as bandwidth estimates, RTOmin, etc.) should be possible, regardless of the underlying transport.  That would enable the application to determine when to send *less* than would be permitted by the underlying transport (e.g. when queues are building).  Dealing with multiplicative increase would be more difficult.  However, HTTP/2 is a failover option (only negotiated when QUIC is not available) so it may not be reasonable to demand optimum performance.

**Comment by murillo128** [comment]
RMCAT algorithms (specially SCREAM) requires a deep integration with the encoder, same happens when doing probing which behaves better if using with FEC data.  So this would be better handled within the js app code. 

However, doing a BWE in js code will always be limited by the underlying QUIC congestion control so won't work either.

Note that to be able to successfully implement an RTC app on js with quictransport we would need the native implementation to provide the instant value for the bandwidth estimation in order to properly change the configuration of the encoder and avoid filling up sending queues that would increase delay. 

I am not familiar with QUIC packet feedback methods, but if they are similar to SCTP, they would be based on congestion and acks windows, which would not be compatible with the current RMCAT algorithms that require a per packet received timing information.

**Comment by aboba** [comment]
@murillo128 There appear to be gaps in both WebTransport and WebCodecs APIs that will make it difficult to implement congestion control in realtime communications applications.  

As noted in the [RTP over QUIC presentation at IETF 111](ttps://datatracker.ietf.org/meeting/111/materials/slides-111-avtcore-ietf-111-avtcore-slides-01), QUIC metrics tend to be less granular (e.g. RTT estimates every few packets), but there is detailed loss info which can be helpful in deciding whether to adjust encoder rate or drop a layer.   So far, attempts to run media over QUIC (e.g. RUSH, RTP over QUIC) appear to have integrated the application with QUIC so as to be able to get access to this information, which is not available in WebTransport.  As an example, when using partial reliability (e.g. reliable streams with a timer) it would be useful to know whether an increase in per-frame delay was due to loss.

An even more challenging question is when to resume sending a layer after it has been dropped. Typically, the BWE will only increase slowly after loss (e.g. additive increase, multiplicative decrease) but with simulcast and SVC, increases tend to be multiplicative. The application can try to continually push the congestion window by adding probes, but it can't exceed the congestion window ramp up built into the QUIC implementation without increasing the send queue. 

As you say, the encoder rate control is done within the application.  Ideally this would allow for per-frame control so as to allow the encoder to quickly ramp up the rate in a non-simulcast/SVC case. However, all that is supported in WebCodecs is an average bitrate target, which can overshoot or undershoot. WebCodecs Issues relating to this include:  [270](https://github.com/w3c/webcodecs/issues/270), [285](https://github.com/w3c/webcodecs/issues/285), [308](https://github.com/w3c/webcodecs/issues/308).

**Comment by martinthomson** [comment]
Right now, it is not possible to do anything along these lines.  Doing anything would be a major project.  It might be possible to have browsers implement a choice of congestion controllers, but that means having browsers reach agreement on what congestion controllers are used.  We might agree to implement one each of a class of congestion controller (throughput-optimized, real-time, whatever) and allow for diversity of approach.  Even that is a major project.

Letting JS implement a congestion controller has a whole host of additional problems.  What metrics it might need, what safeguards need to be in place, and so forth.  I know that real-time applications often do this sort of thing and might even be accustomed to this level of access and control, but working out how to integrate that into a system where the thing being given control is untrustworthy is HARD.

A long time ago, the theory was that real-time applications needed certain amounts of support from the network.  There was lots of effort invested in QoS measures, from the different ATM modes to DSCP.  These were great for research and lots of papers were published.  Then Skype (and others) shipped video calling over the Internet, demonstrating that as long as you have *something*, then you can still deliver an experience that is pretty good.

Maybe sites won't be able to eke out the last iota of performance from the network as a result of not being able to precisely tune every aspect of the browser's network usage to their requirements.  But maybe that will be good enough.  I would need to see better evidence that the baseline thing was not good enough before I would be supportive of investing significant effort into enabling extensive control.

**Comment by aboba** [comment]
I agree - the [WebTransport Use Cases](https://github.com/w3c/webtransport/blob/main/use-cases.md) requiring low-latency streaming from the browser to the server (such as video conferencing) aren't viable. It is still possible to implement low-latency streaming from the cloud to the browser by implementing custom congestion control on the cloud server.  So the Cloud Gaming, AR/VR, and remote desktop scenarios still seem workable. Video ingestion could still make sense if there is a bias towards higher quality versus latency.  But I wonder it is possible to implement something like WISH using the WebTransport API.  This would require uploading simulcast and/or SVC, with the client dynamically adjusting the number of layers being sent.

**Comment by martinthomson** [comment]
There are probably some things that can be done that will deliver a lot of value.  JS-driven congestion control probably isn't in that set.

**Comment by murillo128** [comment]
I think that implementing a bwe algorithm in the js app is completely doable, the main problem is how to bypass the QUIC built-in one.

We can't just disable it for datagrams and trust the js to do the right thing. 

Not sure if it could be possible, but maybe we could relax the congestion control for datagrams and allow the application to send more data than what is allowed by the congestion/ack window without buffering. The quic stack could detect when the js is actually causing congestion by checking the reports 9f the sent packets and apply more tight restrictions.

Another thing that could be done could be to expose an api so the app can hint the quic stack which kind of usage they intend to use and switch between a packet loss based bwe (tcp/vegas/reno...) or a delay one (i.e. bbr/copa...).

**Comment by vasilvv** [comment]
Is this a duplicate of #21?

**Comment by aboba** [comment]
Not a duplicate. Access to BWE and RTT info would help the application to limit buildup of queues by sending less than what might be allowed.  For example, the application could lower the encoder rate, or drop a simulcast or SVC layer in response to increasing delay, before loss is experienced.  This works because increasing delays indicate that the current BWE is approaching the bottleneck bandwidth. 

The more difficult question is when the application can add back a simulcast or SVC layer, a multiplicative increase.  If the application is experiencing low delay and no loss, this doesn't necessarily mean that this would continue after a multiplicative increase.  WebRTC uses probes to more quickly determine when a layer can be added back.

**Comment by jan-ivar** [comment]
I'm hearing some disagreement over what is and isn't possible today. The chairs think this issue would be worth discussing at TPAC.

@aboba would you be able to show your attempts at writing a demo of this, and where exactly you ran into difficulties? More specifically, would you be able to present slides at TPAC to help explain this?

**Comment by jan-ivar** [comment]
Changing milestone for now, as this is going to take a while. Open to further discussion.

**Comment by aboba** [comment]
We are soliciting researchers to present their work at the AVTCORE WG virtual interim in May 2022.

**Comment by notedit** [comment]
In our use case, We want to use WebCodecs + WebTransport + WebAssembly to replace WebRTC. 
We compare the WebTransport version with our WebRTC version. 
WebTransport version can not work very well in 20% packet loss.  Our WebRTC version can work well in 40%+  packet loss. 

So If we can not modify the WebTransport's Congestion Control part,  WebTransport can not compete with WebRTC in Real Time communication.

**Comment by aboba** [comment]
WebTransport is a client/server specification, so it cannot replace WebRTC in P2P scenarios.  Where you control the WebTransport server (e.g. low latency streaming),  you can replace the congestion control with say, Google CC.  This is very commonly done today for RTCDataChannel streaming, which can handle both P2P and client/server scenarios.  This leaves the major broken WebTransport use case as one where the client is sending media to the server.  Is this what you are trying to do?

**Comment by notedit** [comment]
> 

Yes,  We use in SFU Mode not P2P Mode,  We do Congestion Control in server side And send the bitrate to client. But WebTransport(quic) still has his own Congestion Control,  WebTransport may provide a way to disable his own Congestion Control  like  @murillo128 said.

**Comment by notedit** [comment]
> WebTransport is a client/server specification, so it cannot replace WebRTC in P2P scenarios; something more like P2P QUIC would be needed. Where you control the WebTransport server (e.g. low latency streaming), you can replace the congestion control with say, Google CC. This is very commonly done today for RTCDataChannel streaming, which can handle both P2P and client/server scenarios. This leaves the major broken WebTransport use case as one where the client is sending media to the server. Is this what you are trying to do?

Yes, We use in SFU Mode not P2P Mode, We do Congestion Control in server side And send the bitrate to client. But WebTransport(quic) still has his own Congestion Control, WebTransport may provide a way to disable his own Congestion Control like @murillo128 said.

**Comment by jan-ivar** [comment]
We've added the label to discuss at next meeting whether to block Candidate Recommendation on this again.

To keep this simple and spur some discussion: since W3C is primarily concerned with API, how about:
```js
wt.congestionControl = “standard” | “media"
```

**Comment by aboba** [comment]
Maybe express the goal of the CC algorithm? For example "low latency", with "throughput" being the default?

**Comment by jan-ivar** [comment]
Unless changing cc on the fly is important, it should probably be a constructor argument instead:
```js
const wt = new WebTransport(url, {congestionControl: "low-latency"}); // default = "throughput"
```

**Comment by aboba** [comment]
Yes, a constructor argument seems logical.

**Comment by aboba** [comment]
Is there interest in implementing this optional argument (at least as an experiment) so that it can be evaluated?  Some of the comments have argued for a "circuit breaker" mechanism rather than a CC algorithm.  While that would provide considerable freedom for an experiment, I'm not sure that could get approval to ship.

**Comment by jan-ivar** [comment]
Meeting:
- We could put this in a CR without requiring an implementation
- Might need an experiment
- Might need more evidence this would work with the other QUIC pieces
- Some tests have been done but are not conclusive
- A constructor argument is probably sufficient (and a readonly attribute to confirm the dictionary member took)
- Problem: Two streams under the same connection, one uploading video in realtime, the other uploading a large file.
  -  Likely to not work well with two connections, "throughput" one would win. Solving it might involve api's for bandwidth reservation between connections as well.
  - Longing for unified congestion control. Not solved. E.g. RTCDataChannels can choke audio/video in WebRTC if sending anything larger than small messages
  - Connction-level vs. cross-connection. Another option might be: What congestion control is this stream governed by?
- ACTION: Identify groups that would want to experiment in this area.
- Seems like a research project we might need to monitor over time.
- Some have done app-layer cc

**Comment by jan-ivar** [comment]
Meeting:
- David: Can we punt this? Could be added later as an extension?
- Bernard: it's blocking client sending of low-latency media
- Interest in just putting in the API to unblock W3C and rip it out later.

**Comment by bschlinker** [comment]
In addition to allowing a congestion control objective to be passed in (e.g., `low-latency`), can we also consider allowing the congestion control algorithm (CCA) to be disabled outright? I'm interested in this for a few reasons:

**(1) It's necessary to support some types of network conditions measurements.**
- For instance, if I want to evaluate the uplink between a browser and the server, I may want to send datagrams at a fixed rate and observe if and when they arrive at the server to evaluate how network conditions change at short timescales, similar to what an iperf test enables. 
- Pacing packets out in user space is going to inherently limit the fidelity of such measurements, but I suspect it is possible to account for that. However if there's a congestion control algorithm also deciding when the datagrams get sent then it becomes difficult to make sense of the measurements at the remote, especially if the sender in the browser has no insight into when the CCA sent the packet. This is further complicated by the fact that the CCA implementation may vary between WebTransport implementations (intentionally and unintentionally), so trying to account for the CCA is not scalable or tractable. 

**(2) It will enable improvements to the default set of congestion control algorithms.**

- If the standard allows some default set of congestion control options, there's going to naturally be questions about possible improvements. Thus, it's critical that we have a way to execute experiments.
- If it's possible to disable the congestion control algorithm and use a custom algorithm, then it will be possible to  experiment with alternate CCAs at scale. If there's no way to disable the underlying CCA, then work on improvements will be hindered and fractured.
- With WASM + WebTransport, writing a custom congestion control algorithm seems feasible, and it seems possible to write a rudimentary CCA with the current interface. Getting more timestamp information and allowing datagrams to be scheduled would further improve what's possible. I don't think any of this is a pipe dream.

In terms of overhead and risks:
- The current proposal of setting an objective already requires that it be possible to use a different congestion control algorithm for the WebTransport portion of the connection. Thus, being able to outright disable the congestion control used for that portion of the connection would not create an additional burden on implementations.
- Since the destination of all datagram packets is the remote, if the WebTransport use case behaves poorly with the congestion control disabled, the most likely outcome is that it causes the browser's access uplink to become congested. That same problem can be created a number of different ways today by a misbehaving application, so this doesn't seem to be a new risk.

**Comment by jan-ivar** [comment]
> can we also consider allowing the congestion control algorithm (CCA) to be disabled outright

That seems problematic to expose to a single web page, and risks exposing too much information about the user's internet connection and concurrent user activity in other tabs. cc @martinthomson 

To quote our [explainer](https://github.com/w3c/webtransport/blob/main/explainer.md#problem-and-motivation): _"Native applications can use raw UDP sockets, but those are not available on the web because they lack encryption, congestion control, and a mechanism for consent to send (to prevent DDoS attacks)."_

**Comment by martinthomson** [comment]
I think that the non-goals piece covers it adequately.  We can't give websites access to the sort of DoS tool that removing congestion control would provide.

**Comment by aboba** [comment]
It's probably not even realistic to support "circuit-breakers" as the sole congestion control mechanism (e.g. [RFC 8083](https://datatracker.ietf.org/doc/rfc8082/) and [8084](https://datatracker.ietf.org/doc/rfc8084/)).

**Comment by StuartCheshire** [comment]
I fear there is a risk of overcomplicating things here. For a long time people have assumed that network applications can’t have both high throughput and low latency at the same time, so they have to choose one or the other. Recent work is showing this to be false. For example, see Cloudflare’s recent article “[Optimizing TCP for high WAN throughput while preserving low latency](https://blog.cloudflare.com/optimizing-tcp-for-high-throughput-and-low-latency/)”. Also see the Apple developer conference presentation “[Reduce networking delays for a more responsive app](https://developer.apple.com/videos/play/wwdc2022/10078/)”, particularly the part about eliminating excessive unnecessary sender-side buffering. Current work is showing that without fixing sender-side bufferbloat, achieving low latency is at best unreliable, depending more or less randomly on the occupancy of the outgoing buffer. On the other hand, if sender-side bufferbloat is fixed, and the outbound queue is not allowed to grow to excessive size, then all traffic gets low latency.

Similarly, the idea of priority levels was originally rooted in an assumption of bandwidth scarcity. There’s an assumption that for one traffic flow to get the capacity it needs, some other traffic flow must be made to suffer. Commonly today we have enough bandwidth to do what the user is trying to do, and priority levels function primarily as a way to decide which packets get to skip to the front of an excessively large queue. This is only a benefit if we believe that having an excessively large output queue is inevitable. Priority levels serve to manage delay in a bloated queue, not to allocate scarce shares of inadequate bandwidth. If we fix the sender-side bufferbloat so that _all_ packets experience low delay, then we no longer have to prioritize which packets deserve low delay and which do not.

**Comment by jan-ivar** [comment]
Meeting:
- Hearing more interest in more abstract API or no API than the API that exposes browser's congestion control algorithms by name.
- To move forward, go back to initial PR with "default" (instead of "throughput") and "low-latency" hints only that don't throw (with readable attribute to verify setter only), and then mark it as Feature at Risk

**Comment by kixelated** [comment]
Hey @StuartCheshire, I completely agree that congestion control and prioritization are only applicable when bandwidth limited, and only relevant when an application wants to minimize latency. Live media is the classic use-case where both are needed.

You strike the nail on the head when you say that "low-latency" is all about queue management. In an ideal world, there would be no queues between the sender and the receiver, however layers get in the way. Libraries will queue, sockets will queue, intermediaries will queue, and even hardware will queue.

The application can detect and react to queueing. That's the appeal of UDP, although then each application has their own bespoke method of congestion control, fragmentation, and flow control.

For the sake of the internet and ecosystem, I appreciate that WebTransport enforces congestion control. I think a hint is a good idea, but it's still required to address _why_ applications implemented their own congestion control in the first place.

In the case of live media, we have tight latency requirements, but we're also application-limited and need relatively constant throughput. A "low-latency" congestion control algorithm like BBR is not necessarily a good fit because the rate estimation does not function when application-limited and the PROBE_RTT phase significantly reduces throughput for ~300ms at frequent intervals 

That being said, I still think the hints are the best approach. Ultimately you're giving the application control some control over the trade-offs in a diverse congestion control ecosystem. I don't care if GCC is used, but I do care that something with similar properties is used when I hint "real-time media".


As for priorities, I don't want to derail this conversation too much but it comes back to reducing queuing, although this time in the application -> library interaction.

The classic TCP approach to avoid queues in the network socket is low watermark. Basically the application keeps a priority queue in memory, writes juuuust enough data to saturate the socket queue, and schedules a wakeup when the socket queue size drops below a certain threshold. The threshold dictates the latency versus efficiency trade-off. This approach is fine and could be employed in WebTransport.

I think a better approach is to provide the library with the priority queue itself. This fits very nicely with QUIC's API of independent streams and avoids the latency versus efficiency trade-off. Effectively the application gives WebTransport simple rules on which packet/stream to send next to avoid any additional communication between the browser and JS layers.

The trick is making sure the prioritization rules are simple enough (ex. not H/2). Warp uses a numeric priority like a classic priority queue, where higher valued streams are transmitted first. This won't work for every use-case but I think it covers a lot of ground without adding much complexity.

**Comment by kixelated** [comment]
One more point:

>  For a long time people have assumed that network applications can’t have both high throughput and low latency at the same time, so they have to choose one or the other. Recent work is showing this to be false.

I want to point out that network latency is separate from application latency. I don't actually care how long it takes to transmit a packet; I only care about how long it takes for my encoded frame to reach the destination. During congestion, the application is bandwidth limited, and that causes application latency if nothing is done to mitigate it, regardless of if the queue is over the network (bufferbloat) or within my application.

A low latency application needs to adjust its bitrate in response to the network bitrate. For live media, this means dropping the required bandwidth by changing the encoding bitrate, choosing another rendition, or delay/drop frames altogether. Low network latency allows the application to make these decisions quickly, provided there are APIs available to prioritize or get notified of large queue sizes.

**Comment by fideltian** [comment]
we are very interested on the topic.  It seems the topic has been closed. Any conclusion?
our case is :  transform realtime data on Webtransport. But seems congestion control within webtranport(uplink) will have side effect, comparing with RTP congestion control.
